Local-mode build may fail in certain Go configurations

A recent case indicates that local-mode build via `make` fails in a system:
- Ubuntu 18.04
- Go 1.11.5 installed using available-on-net instructions, such that it ends up in /usr/local
  (note that distro-supported newest Go is 1.10.4 and gets installed in /usr/lib, and will not expose this issue)
- libndctl newest v64.1 built manually with default ./configure flags and installed (shared libs go to /usr/lib/)
  (note that distro-supported newest libndctl is v61)

On such a system, make will fail like this:
```
GOOS=linux go build -a -o _output/pmem-csi-driver ./cmd/pmem-csi-driver # github.com/intel/pmem-csi/pkg/ndctl
/tmp/go-build681207083/b159/_x007.o: In function `_cgo_0eba3381006a_Cfunc_ndctl_region_get_max_available_extent':
/tmp/go-build/cgo-gcc-prolog:264: undefined reference to `ndctl_region_get_max_available_extent'
collect2: error: ld returned 1 exit status
Makefile:52: recipe for target 'pmem-csi-driver' failed
```

--- a/DEVELOPMENT.md
+++ b/DEVELOPMENT.md
@@ -1,0 +1,256 @@
+Table of Contents
+----------------
+
+- [Code quality](#code-quality)
+- [APIs](#apis)
+- [Development mode build and run](#development-mode-build-and-run)
+- [Notes about switching DeviceMode](#notes-about-switching-devicemode)
+- [Notes about accessing system directories in a container](#notes-about-accessing-system-directories-in-a-container)
+
+Code quality
+============
+
+Coding style
+------------
+
+The normal Go style guide applies. It is enforced by `make test`, which calls `gofmt`. `gofmt` should be
+from Go >= v1.11.4, older version (like v1.10) format the code slightly differently.
+
+
+Input validation
+----------------
+
+In all cases, input comes from a trusted source because network
+communication is protected by mutual TLS and the `kubectl` binaries
+runs with the same privileges as the user invoking it.
+
+Nonetheless, input needs to be validated to catch mistakes:
+
+- detect incorrect parameters for `kubectl`
+- ensure that messages passed to gRPC API implementations
+  in registry, controller and driver have all required
+  fields
+- the gRPC implementation rejects incoming messages that are too large
+  (https://godoc.org/google.golang.org/grpc#MaxRecvMsgSize) and
+  refuses to send messages that are larger
+  (https://godoc.org/google.golang.org/grpc#MaxSendMsgSize)
+
+
+APIs
+============
+
+​CSI API
+----------------
+Kubernetes CSI API is exposed over Unix domain socket. CSI operations are executed as gRPC calls. Input data is allowed as permitted by CSI specification. Output data is formatted as gRPC response.
+
+Following CSI operations are supported, with arguments as specified by CSI specification: CreateVolume, DeleteVolume, StageVolume, UnstageVolume, PublishVolume, UnpublishVolume, ListVolumes, GetCapacity, GetCapabilities, GetPluginInfo, GetPluginCapabilities.
+
+
+Network ports
+-------------
+
+Network ports are opened as configured in manifest files:
+
+- registry endpoint: typical port value 10000, used for pmem-csi internal communication
+- controller endpoint: typical port value 10001, used for serving CSI API
+
+
+Local sockets
+-------------
+Kubernetes CSI API used over local socket inside same host.
+
+- unix:///var/lib/kubelet/plugins/pmem-csi-reg.sock
+- unix:///var/lib/kubelet/plugins/pmem-csi/csi.sock
+- unix:///var/lib/kubelet/plugins/pmem-csi/csi-controller.sock
+
+
+Command line arguments
+----------------------
+Note that different set of programs is used in different DeviceModes. Three stages: *pmem-ns-init*, *pmem-vgm*, *pmem-csi-driver* run in DeviceMode:LVM. Only *pmem-csi-driver* runs in DeviceMode:Direct.
+
+Common arguments
+----------------
+
+argument name            | meaning                                           | type   | range
+-------------------------|---------------------------------------------------|--------|---
+-alsologtostderr         | log to standard error as well as files            |        |
+-log_backtrace_at value  | when logging hits line file:N, emit a stack trace |        |
+-log_dir string          | If non-empty, write log files in this directory   | string |
+-log_file string         | If non-empty, use this log file                   | string |
+-logtostderr             | log to standard error instead of files            |        |
+-skip_headers            | avoid header prefixes in the log messages         |        |
+-stderrthreshold value   | logs at or above this threshold go to stderr (default 2) | |
+-v value                 | log level for V logs                              | int    |
+-vmodule value           | comma-separated list of pattern=N settings for file-filtered logging | string |
+
+Specific arguments to pmem-ns-init
+-----------------------------------
+
+argument name      | meaning                                                | type | range
+-------------------|--------------------------------------------------------|------|---
+-namespacesize in  | Namespace size in GB (default 32)                      | int  | set to 2 if less than 2
+-useforfsdax int   | Percentage of total to use in Fsdax mode (default 100) | int  | 0..100
+-useforsector int  | Percentage of total to use in Sector mode (default 0)  | int  | 0..100
+
+Note: useforfsdax + useforsector must be <=100
+
+
+Specific arguments to pmem-vgm
+-------------------------------
+NO SPECIFIC arguments
+
+Specific arguments to pmem-csi-driver
+--------------------------------------
+
+argument name        | meaning                                           | type | allowed | default
+---------------------|---------------------------------------------------|------|---------|---
+-caFile string       | Root CA certificate file to use for verifying connections | string | |
+-certFile string     | SSL certificate file to use for authenticating client connections(RegistryServer/NodeControllerServer) | string | |
+-clientCertFile string | Client SSL certificate file to use for authenticating peer connections | string | | certFile
+-clientKeyFile string | Client private key associated to client certificate | string |              | keyFile
+-controllerEndpoint string | internal node controller endpoint              | string |              |
+-deviceManager string      | device manager to use to manage pmem devices   | string | lvm or ndctl | lvm
+-drivername string         | name of the driver                             | string |              | pmem-csi
+-endpoint string           | PMEM CSI endpoint                              | string |              | unix:///tmp/pmem-csi.sock
+-keyFile string            | Private key file associated to certificate     | string |              |
+-mode string               | driver run mode                                | string | controller, node, unified | unified
+-nodeid string             | node id                                        | string |              | nodeid
+-registryEndpoint string   | endpoint to connect/listen registry server     | string |              |
+
+
+Environment variables
+---------------------
+
+TEST_WORK is used by registry server unit-test code to specify path to certificates in test system. 
+Note, THIS IS NOT USED IN PRODUCTION
+
+Development mode build and run
+==============================
+
+It is possible to build and try the plugin on a stand-alone system without Kubernetes involved. This may be useful in the development mode. The plugin can be started in Unified mode where both Controller and Node functions are served by a single process running on one host.
+
+Development system using Virtual Machine with emulated devices
+--------------------------------------------------------------
+
+Development and some verification can be performed using QEMU-emulated persistent memory devices instead of HW-based persistent memory devices. See [README-qemu-notes](README-qemu-notes.md) for more details about VM configuration.
+
+
+Build plugin for stand-alone use
+--------------------------------
+
+Building locally for stand-alone use has been verified using these components:
+
+- Go: versions 1.10, 1.11 are usable for building, go 1.11 is required for clean run of 'make test`
+- [ndctl](https://github.com/pmem/ndctl) versions 62..64, either built on dev.host via autogen, configure, make, and installed using instructions in README.md, or installed as ndctl package(s) from distribution repository.
+
+Use `make`
+
+This produces the following binaries in the `_output` directory:
+
+  * `pmem-ns-init`: Utility for namespace initialization in DeviceMode:LVM
+  * `pmem-vgm`: Utility for creating logical volume groups over PMEM devices created, used in DeviceMode:LVM
+  * `pmem-csi-driver`: PMEM-CSI driver, used in both DeviceModes
+
+
+Stand-alone mode build potential issues
+---------------------------------------
+
+In a recent case study it was observed that stand-alone mode build may fail:
+
+- Ubuntu 18.04
+- Go 1.11.5 installed into /usr/local using available-on-net instructions
+  (note that distro-supported Go is 1.10.4 and gets installed in /usr/lib, and will not expose this issue)
+- libndctl v64.1 built manually using `./configure` and installed, with shared libs end up in /usr/lib/
+- note that distro-supported libndctl would be v61
+
+On such a system, stand-alone mode make will fail like this:
+```
+GOOS=linux go build -a -o _output/pmem-csi-driver ./cmd/pmem-csi-driver # github.com/intel/pmem-csi/pkg/ndctl
+/tmp/go-build681207083/b159/_x007.o: In function `_cgo_0eba3381006a_Cfunc_ndctl_region_get_max_available_extent':
+/tmp/go-build/cgo-gcc-prolog:264: undefined reference to `ndctl_region_get_max_available_extent'
+collect2: error: ld returned 1 exit status
+Makefile:52: recipe for target 'pmem-csi-driver' failed
+```
+
+The problem seems to be that CGo linker looks for shared libraries in GOROOT-related directories, and does not consider /usr/lib.
+One work-around is to prepend linker flags to compile command by replacing line in Makefile:
+
+`GOOS=linux go build -a -o _output/$@ ./cmd/$@`
+
+with line:
+
+`CGO_LDFLAGS=-L/usr/lib GOOS=linux go build -a -o _output/$@ ./cmd/$@`
+
+This example demonstrates that both Go environment and libndctl shared libraries may be built and/or installed manually on various paths, bypassing package management, as newer-than-distro-provided versions are sometimes desired. It is not straightforward to support all different combinations without creating complexity-adding dynamic detection of components, but that is out of scope currently, so some local work-arounds may be needed.
+
+Run driver as stand-alone program
+---------------------------------
+
+Use `util/run-lvm-unified` as user:root to start the driver in DeviceMode:LVM.
+This runs two preparation parts and starts the driver binary, which listens and responds to API use on a TCP socket. You can modify this to use a Unix domain socket, if needed.
+Use `util/run-direct-unified` as user:root to start the driver in DeviceMode:direct. In this mode the two preparation stages are skipped.
+
+Verify driver in unified mode
+-----------------------------
+
+Some of the functionality of the driver can be verified in the single-host context. This running mode is called "Unified" in the driver. Both Controller and Node service run combined in local host, without Kubernetes context.
+
+The endpoint for driver access can be specified either:
+
+* with each csc command as `--endpoint tcp://127.0.0.1:10000`
+* export endpoint as env.variable, see `util/lifecycle-unified.sh`
+
+These run-time dependencies are used by the plugin in the Unified mode: lvm2 shred mount file blkid
+
+Scripts in util/ directory
+--------------------------
+
+* [lifecycle-unified](util/lifecycle-unified.sh) example steps verifying a volume lifecycle
+* [sanity-unified](util/sanity-unified.sh) API test using csi-sanity
+* [get-capabilities-unified](util/get-capabilities-unified.sh) Query Controller and Node capabilities
+
+These utilities are required by scripts residing in `util/` directory:
+
+- [csc](http://github.com/rexray/gocsi)
+- [csi-sanity](http://github.com/kubernetes-csi/csi-test)
+
+Override registry address
+-------------------------
+Sometimes images need to be deployed from specific registry instead of what is written in deployment files. You can use this example command to change registry address in all manifest files under deploy/ directory:
+```
+for f in $(find ./deploy -type f); do sed -i -e 's/192.168.8.1:5000/MY_REGISTRY/' $f; done
+```
+
+Notes about switching DeviceMode
+================================
+
+If DeviceMode is switched between LVM and Direct(ndctl), please keep in mind that pmem-csi driver does not clean up or reclaim Namespaces, therefore Namespaces plus other related context (possibly LVM state) created in previous mode will remain stored on device and most likely will create trouble in another DeviceMode.
+
+Going from DeviceMode:LVM to DeviceMode:Direct
+----------------------------------------------
+- examine LV Groups state on a node: `vgs`
+- examine LV Phys.Volumes state on a node: `pvs`
+- Delete LV Groups before deleting namespaces: `vgremove VGNAME`, to avoid orphaned VGroups
+
+NOTE: The next **WILL DELETE ALL NAMESPACES** so be careful!
+- Delete Namespaces on a node using CLI: `ndctl destroy-namespace all --force`
+
+Going from DeviceMode:Direct to DeviceMode:LVM
+----------------------------------------------
+No special steps are needed to clean up Namespaces state.
+
+If pmem-csi driver has been operating correctly, there should not be existing Namespaces as CSI Volume lifecycle should have been deleted those after end of life of Volume. If there are, you can either keep those (DeviceMode:LVM does honor "foreign" Namespaces and leaves those alone) if you have enough space, or you can choose to delete those using `ndctl` on node.
+
+Notes about accessing system directories in a container
+=======================================================
+
+The pmem-csi driver will run as container, but it needs access to system directories /sys and /dev. Two related potential problems have been diagnosed so far.
+
+Read-only access to /sys
+------------------------
+In some deployment schemes /sys remains mounted read-only in the container running pmsm-csi-driver. This is not problem in DeviceMode:LVM, but is a blocking problem in DeviceMode:Direct where the driver needs write access to /sys for Namespaces management operations. There is start-time check for read-write mount of /sys in the code. An error in pod log `pmem-driver: Failed to run driver: FATAL: /sys mounted read-only, can not operate` is the sign of such state.
+
+Access to /dev of host
+----------------------
+Containers runtime may not pass /dev from host into the container. This is, again, problem in DeviceMode:Direct. If the /dev/ of the host is not accessible in the pmem-csi container, there will be failure in accessing of newly created block device /dev/pmemX.Y which will not be visible inside container. The driver does not detect the root cause of that problem during start-up, but only when a volume creation has failed. This problem can be avoided by specifying explicit mount of /dev in the pmem-csi manifest. 
+

--- a/README-qemu-notes.md
+++ b/README-qemu-notes.md
@@ -1,10 +1,8 @@
 # Notes about VM config for Pmem-CSI development environment
 
-
-> **About hugepages:**
-> Emulated NVDIMM appears not fully working if the VM is configured to use Hugepages. With Hugepages configured, no data survives guest reboots and nothing is ever written into backing store file in host. Configured backing store file is not even part of command line options to qemu. Instead of that, there is some dev/hugepages/libvirt/... path which in reality remains empty in host.
-
-These are some notes documenting VM configuration for pmem-csi development using emulated NVDIMM as host-backed files. For more choices of VM setup, please also check the test/ directory where there is automated method of VM creation implemented.
+These are some notes about manually created VM configuration for pmem-csi development using emulated NVDIMM as host-backed files.
+There exists a newer, more convenient automated method using code in test/ directory, where a four node Kubernetes cluster
+can be create by simply typing `make start`.
 
 VM configuration described here was used in early pmem-csi development where a VM was manually created and then used as development host. The initial VM config was created by libvirt/GUI (also doable using virt-install CLI), with some configuration changes made directly in VM-config xml file to emulate a NVDIMM device backed by host file. Two emulated NVDIMMs were tried at some point, but operations on namespaces appear to be more reliable with single emulated NVDIMM.
 
@@ -56,7 +54,7 @@ Another, quicker way of creating such file is to use `truncate -s` which creates
 truncate -s 16G /var/lib/libvirt/images/nvdimm0
 ```
 
-qemu, if using libvirt-created config will then that file with zeroes, as it uses prealloc=yes` attribute for backing file
+qemu using libvirt-created configuration will fill that file with zeroes, as it uses `prealloc=yes` attribute for backing file.
 
 ## Labels initialization is needed once per emulated NVDIMM
 
@@ -70,3 +68,7 @@ ndctl init-labels nmem0
 ndctl enable-region region0
 
 ```
+
+## Note about hugepages
+
+Emulated NVDIMM appears not fully working if the VM is configured to use Hugepages. With Hugepages configured, no data survives guest reboots and nothing is ever written into backing store file in host. Configured backing store file is not even part of command line options to qemu. Instead of that, there is some dev/hugepages/libvirt/... path which in reality remains empty in host.

--- a/README.md
+++ b/README.md
@@ -171,14 +171,7 @@ Volumes of this type shall be used in combination with `volumeBindingMode: Immed
 
 ### Software required
 
-Building has been verified using these components:
-
-- Go: version 1.10.1 (go 1.11 is required for 'make test`)
-- [ndctl](https://github.com/pmem/ndctl) versions 62..64, either built on dev.host via autogen, configure, make, and install as per instruction in README.md, or installed as ndctl package(s) from distribution repository.
-
-Building of Docker images has an additional requirement:
-
-- Docker-ce: verified using version 18.06.1
+Building of Docker images has been verified using Docker-ce: version 18.06.1
 
 ### Hardware required
 
@@ -198,48 +191,25 @@ PMEM-CSI driver implements CSI specification version 1.0.0, which only supported
 
 ## Setup
 
-### Development system using Virtual Machine
-
-Early development and verification was performed on QEMU-emulated persistent memory devices.
-
-The build was verified on the system described below:
-
-* Host: Dell Poweredge R620, distro: openSUSE Leap 15.0, kernel 4.12.14, qemu 2.11.2
-* Guest VM: 32GB RAM, 8 vCPUs, Ubuntu 18.04.1 server, kernel 4.15.0, 4.18.0, 4.19.1
-* See [README-qemu-notes](README-qemu-notes.md) for more details about VM config
-
 ### Get source code
 
-Use the command: `git clone https://github.com/intel/pmem-csi`
+Use these commands:
+
+```
+mkdir -p $GOPATH/src/github.com/intel
+git clone https://github.com/intel/pmem-csi $GOPATH/src/github.com/intel/pmem-csi
+```
 
 ### Build plugin
 
-1.  Use `make`
+1.  Use `make build-images` to produce Docker container images.
 
-    This produces the following binaries in the `_output` directory:
-
-    * `pmem-ns-init`: Helper utility for namespace initialization.
-    * `pmem-vgm`: Helper utility for creating logical volume groups over PMEM devices created.
-    * `pmem-csi-driver`: PMEM-CSI driver.
-
-2.  Use `make build-images` to produce Docker container images.
-
-3.  Use `make push-images` to push Docker container images to a Docker images registry. The
+2.  Use `make push-images` to push Docker container images to a Docker images registry. The
     default is to push to a local [Docker registry](https://docs.docker.com/registry/deploying/).
 
 See the [Makefile](Makefile) for additional make targets and possible make variables.
 
 ### Run plugin
-
-#### Run driver as standalone
-
-This is useful in development/trial mode.
-
-Use `util/run-lvm-unified` as user:root.
-This runs two preparation parts, and starts the driver binary, which listens and responds to API use on a TCP socket. You can modify this to use a Unix socket, if needed.
-Use `util/run-direct-unified` as user:root to start the driver in DeviceMode:direct. This script skips the two preparation stages and starts the driver binary with corresponding DeviceMode option.
-
-#### Run as Kubernetes deployment
 
 This section assumes that a Kubernetes cluster is already available
 with at least one node that has persistent memory device(s). For development or
@@ -254,26 +224,23 @@ machines, see the ["End-to-end testing"](#end-to-end-testing) section below.
 
 The label **storage: pmem** needs to be added to the cluster node that provides persistent memory device(s).
 
-Clusters with multiple nodes with persistent memory are not fully supported at
-the moment. Support for this will be added when making the CSI driver topology-
-aware.
-
-- **Deploy the driver to Kubernetes using DeviceMode:LVM **
+- **Deploy the driver to Kubernetes using DeviceMode:LVM**
 
 ```sh
-    $ sed -e 's/192.168.8.1:5000/<your registry>/' deploy/kubernetes-1.12/pmem-csi-lvm.yaml | kubectl create -f -
+    $ sed -e 's/192.168.8.1:5000/<your registry>/' deploy/kubernetes-<kubernetes version>/pmem-csi-lvm.yaml | kubectl create -f -
 ```
 
-- **Deploy the driver to Kubernetes using DeviceMode:Direct **
+- **Alternatively, deploy the driver to Kubernetes using DeviceMode:Direct**
 
 ```sh
-    $ sed -e 's/192.168.8.1:5000/<your registry>/' deploy/kubernetes-1.12/pmem-csi-direct.yaml | kubectl create -f -
+    $ sed -e 's/192.168.8.1:5000/<your registry>/' deploy/kubernetes-<kubernetes version>/pmem-csi-direct.yaml | kubectl create -f -
 ```
 
 The deployment yaml file uses the registry address for the QEMU test cluster
 setup (see below). When deploying on a real cluster, some registry
 that can be accessed by that cluster has to be used.
-
+If the Docker registry runs on the local development
+host, then the `sed` command which replaces the Docker registry is not needed.
 The `deploy` directory contains one directory or symlink for each
 tested Kubernetes release. The most recent one might also work on
 future, currently untested releases.
@@ -343,35 +310,6 @@ Use the `make test` command.
 
 **Note:** Testing code is not completed yet. Currently it runs some passes using `gofmt, go vet`.
 
-### Verify driver in unified mode
-
-The driver can be verified in the single-host context. This running mode is called "Unified"
-in the driver. Both Controller and Node service run combined in local host, without Kubernetes context.
-
-The endpoint for driver access can be specified either:
-
-* with each csc command as `--endpoint tcp://127.0.0.1:10000`
-* export endpoint as env.variable, see `util/lifecycle-unified.sh`
-
-These run-time dependencies are used by the plugin in Unified mode:
-
-- lvm2
-- shred
-- mount
-- file
-- blkid
-
-#### Scripts in util/ directory
-
-* [lifecycle-unified](util/lifecycle-unified.sh) example steps verifying a volume lifecycle
-* [sanity-unified](util/sanity-unified.sh) API test using csi-sanity
-* [get-capabilities-unified](util/get-capabilities-unified.sh) Query Controller and Node capabilities
-
-These utilities are required by scripts residing in `util/` directory:
-
-- [csc](http://github.com/rexray/gocsi) v0.5.0
-- [csi-sanity](http://github.com/kubernetes-csi/csi-test) v0.2.0-1-95-g3bc4135
-
 ### End-to-end testing (E2E)
 
 #### QEMU + Kubernetes
@@ -380,7 +318,7 @@ E2E testing relies on a cluster running inside multiple QEMU virtual
 machines. This is known to work on a Linux development host system.
 The `qemu-system-x86_64` binary must be installed, either from
 [upstream QEMU](https://www.qemu.org/) or the Linux distribution.
-
+The user must be able to run commands as root via `sudo`.
 For networking, the `ip` tool from the `iproute2` package must be
 installed. The following command must be run once after booting the
 host machine and before starting the virtual machine:
@@ -450,19 +388,24 @@ version supported by Clear Linux installed.
 `make start` will bring up a Kubernetes test cluster inside four QEMU
 virtual machines. It can be called multiple times in a row and will
 attempt to bring up missing pieces each time it is invoked.
-
-Once it completes, everything is ready for interactive use via
+The first node `host-0` is the Kubernetes master without persistent memory.
+The other three nodes are worker nodes with one emulated 32GB NVDIMM each.
+After the cluster has been formed, `make start` adds `storage=pmem` label
+to the worker nodes and deploys the pmem-csi driver.
+Once `make start` completes, the cluster is ready for interactive use via
 `kubectl` inside the virtual machine. Alternatively, you can also
 set `KUBECONFIG` as shown at the end of the `make start` output
-and use a local `kubectl` binary.
+and use `kubectl` binary on the host running VMs.
 
-The first node is the Kubernetes master without persistent memory. The other
-three nodes are worker nodes with one 32GB NVDIMM each. The worker nodes have
-already been labeled with `storage=pmem`, but the pmem-csi driver still needs to be installed manually as shown in ["Run as Kubernetes deployment"](#run-as-
-kubernetes-deployment). If the Docker registry runs on the local development
-host, then the `sed` command which replaces the Docker registry is not needed.
+Use `make stop` to stop the virtual machines. The cluster state remains preserved and will be restored after next `make start`.
 
-Once done, `make stop` will clean up the cluster and shut everything down.
+#### Running commands on test cluster nodes over ssh
+
+`make start` generates ssh-wrappers `_work/ssh-clear-kvm.N` for each test cluster node which are handy for running a single command or to start interactive shell. Examples:
+
+`_work/ssh-clear-kvm.0 kubectl get pods` runs a kubectl command on node-0 which is cluster master.
+
+`_work/ssh-clear-kvm.1` starts a shell on node-1.
 
 #### Running E2E tests
 
@@ -471,9 +414,6 @@ Once done, `make stop` will clean up the cluster and shut everything down.
 tests and some
 [Kubernetes storage tests](https://github.com/kubernetes/kubernetes/tree/master/test/e2e/storage/testsuites)
 against the pmem-csi driver.
-
-The driver will get deployed automatically and thus must not be
-installed yet on the cluster.
 
 When [ginkgo](https://onsi.github.io/ginkgo/) is installed, then it
 can be used to run individual tests and to control additional aspects

